File: segmentManager/src/main/java/com/clumob/segment/manager/SegmentNavigation.java
package com.clumob.segment.manager;

import com.clumob.segment.controller.SegmentInfo;

import java.util.Deque;
import java.util.LinkedList;

public class SegmentNavigation {

    private final SegmentManager segmentManager;
    private Deque<SegmentInfo> backStack = new LinkedList<>();

    public SegmentNavigation(SegmentManager segmentManager) {
        this.segmentManager = segmentManager;
    }

    public void addToBackStack(SegmentInfo segmentInfo) {
        SegmentInfo segmentInfoOld = navigateToScreen(segmentInfo);
        if (segmentInfoOld != null) {
            backStack.add(segmentInfoOld);
        }
    }

    public SegmentInfo navigateToScreen(SegmentInfo<?, ?> segmentInfo) {
        return segmentManager.changeSegment(segmentInfo);
    }

    public boolean popBackStack() {
        boolean b = this.backStack.size() > 0;
        if (b) {
            SegmentInfo segmentInfo = backStack.pollLast();
            navigateToScreen(segmentInfo);
        }
        return b;
    }

    public void clearStack() {
        backStack.clear();
    }

    public int getBackStackSize() {
        return backStack.size();
    }


}

File: segmentManager/src/main/java/com/clumob/segment/manager/SegmentNavigation.kt
package com.clumob.segment.manager

import com.clumob.segment.controller.SegmentInfo
import java.util.*

open class SegmentNavigation(val segmentManager: SegmentManager) {
    private val backStack: Deque<SegmentInfo> = LinkedList()
    open fun addToBackStack(segmentInfo: SegmentInfo) {
        val segmentInfoOld = navigateToScreen(segmentInfo)
        if (segmentInfoOld != null) {
            backStack.add(segmentInfoOld)
        }
    }

    open fun navigateToScreen(segmentInfo: SegmentInfo): SegmentInfo? {
        return segmentManager.changeSegment(segmentInfo)
    }

    open fun popBackStack(): Boolean {
        val b = backStack.size > 0
        if (b) {
            val segmentInfo = backStack.pollLast()
            navigateToScreen(segmentInfo)
        }
        return b
    }

    open fun clearStack() {
        backStack.clear()
    }

    open val backStackSize: Int
        get() = backStack.size

}